### PR TITLE
Denoiser misc work

### PR DESCRIPTION
**Pull request by est77**
- Added basic logging support. 
- Recovered error messages from the original code.
- Fixed bugs when number of samples is 1 and other special cases.
- Added mark invalid pixels diagnostic mode.
- Formatting and code correctness improvements.

### CODE REVIEW

File: src/thirdparty/bcd/bcd/Denoiser.cpp [hunk 1]
```diff
@@ -35,19 +34,21 @@ using namespace boost;
 namespace bcd
 {
 
-bool Denoiser::denoise()
+Denoiser::Denoiser()
+  : IDenoiser()
```
Comment (dictoon) [inline src/thirdparty/bcd/bcd/Denoiser.cpp]: The whole constructor looks redundant.

File: src/thirdparty/bcd/bcd/DenoisingUnit.cpp [hunk 2]
```diff
@@ -120,22 +126,25 @@ void DenoisingUnit::selectSimilarPatches()
         m_searchWindowRadius,
         m_patchRadius);
 
-    m_similarPatchesCenters.clear();
-    m_similarPatchesCenters.reserve(m_maxNbOfSimilarPatches);
+    m_nbOfSimilarPatches = 0 ;
```
Comment (dictoon) [inline src/thirdparty/bcd/bcd/DenoisingUnit.cpp]: Replace space before `;`

File: src/thirdparty/bcd/bcd/DenoisingUnit.cpp [hunk 3]
```diff
@@ -157,29 +166,33 @@ inline float DenoisingUnit::histogramPatchDistance(
         totalNbOfNonBoth0Bins += nbOfNonBoth0Bins;
     }
 
+    // If the histograms have no bins in common, don't consider them close.
+    if (totalNbOfNonBoth0Bins == 0)
+        return std::numeric_limits<float>::max();
```
Comment (dictoon) [inline src/thirdparty/bcd/bcd/DenoisingUnit.cpp]: I think the `std::` qualifier is redundant here.

File: src/thirdparty/bcd/bcd/IDenoiser.cpp [hunk 4]
```diff
@@ -12,81 +12,230 @@
...
 }
 
-IProgressReporter::~IProgressReporter()
+ICallbacks::LogHelper::~LogHelper()
 {
+    string msg = m_ss.str();
+
+    if (!msg.empty())
+    {
+        m_logFn( msg.c_str() ) ;
```
Comment (dictoon) [inline src/thirdparty/bcd/bcd/IDenoiser.cpp]: Remove extra spaces?

File: src/thirdparty/bcd/bcd/Utils.h [hunk 5]
```diff
@@ -26,6 +30,15 @@
 namespace bcd
 {
 
+inline bool isFinite(const float x)
+{
+#ifdef _WIN32
+    return _finite(x) != 0;
+#else
+    return !(std::isnan(x) || std::isinf(x));
```
Comment (dictoon) [inline src/thirdparty/bcd/bcd/Utils.h]: I'm afraid that won't compile with VS 2012.

Comment (est77) [reply src/thirdparty/bcd/bcd/Utils.h]: On VS 2012 we use _finite(). We tested with sergo before that VS2012 supported it.

**Comment by est77** [comment]
Addressed comments and rebased.

**Comment by dictoon** [comment]
Thanks, you can merge when the build passes.

**Pull request merged**